feat: substance program synthesizer

wodeni opened the pull request:
# Description

Related issue/PR: #292 

This PR contains the Substance program synthesizer, which contains a `synthesis` module in `core` and a client application `@penrose/synthesizer`. Both have README files the describe their functionality and usage. 

# Implementation strategy and design decisions

- Merged `penrose-artifacts` into `@penrose/automator`. The automator now serves three purposes: (1) batch-render a registry of Penrose trios, (2) generate metadata such as running time and cross-instance energy for profiling, and (3) generate a static site for previewing the results.
- There are two possible ways to trigger the static site generation: `yarn start batch --render=<folder>` or `yarn start render <artifacts-folder> <browser-folder>`
- The synthesizer now generates a registry of the synthesized files. The first entry in `trios` is the reference/template Substance program.
# Examples with steps to reproduce them

Usage string for synthesizer:
```
Penrose Synthesizer.

Usage:
  penrose-synthesizer <domain> [options]

Options:
  --substance=SUB      A Substance program that will be included in every synthesized program     
  --style=STY          A Style program that the synthesized programs work with
  --path=PATH          Output path for the generated programs (Trailing slash not needed)
                       [default: synthesized-progs]
  --registry=REGISTRY  Path for the generated registry
                       [default: registry.json]
  --num-programs=NUM   The number of programs to generate 
                       [default: 1]
  --synth-setting=SET  A JSON file containing parameters for the synthesizer
```

Usage string for automator:
```
Penrose Automator.

Usage:
  automator batch LIB OUTFOLDER [--folders]  [--src-prefix=PREFIX] [--repeat=TIMES] [--render=OUTFOLDER]
  automator render ARTIFACTSFOLDER OUTFOLDER

Options:
  -o, --outFile PATH Path to either an SVG file or a folder, depending on the value of --folders. [default: output.svg]
  --folders Include metadata about each output diagram. If enabled, outFile has to be a path to a folder.
  --src-prefix PREFIX the prefix to SUBSTANCE, STYLE, and DOMAIN, or the library equivalent in batch mode. No trailing "/" required. [default: .]
  --repeat TIMES the number of instances
```


In `packages/synthesizer`, run:

`yarn start __tests__/setTheory.dsl --substance=__tests__/template.sub  --style=__tests__/venn.sty --path=../automator/progs --num-programs=10`

```
Set A
Set B
Set C
IsSubset(B, A)
Equal(C, B)
Equal(A, C)
Equal(B, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(C, B)
Add Equal(A, C)
Delete IsSubset(C, A)
Add Equal(B, A)
----------
Set A
Set B
Set C
Equal(C, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Delete IsSubset(B, A)
Delete IsSubset(C, A)
Add Equal(C, A)
----------
Set A
Set B
Set C
IsSubset(B, A)
Equal(B, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(B, A)
Delete IsSubset(C, A)
----------
Set A
Set B
Set C
IsSubset(B, A)
Equal(C, B)
Equal(B, B)
Equal(A, C)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(C, B)
Delete IsSubset(C, A)
Add Equal(B, B)
Add Equal(A, C)
----------
Set A
Set B
Set C
Equal(C, C)
Equal(A, C)
ℹ Operations:                                                substance synthesizer 14:07:13
 Delete IsSubset(B, A)
Add Equal(C, C)
Add Equal(A, C)
Delete IsSubset(C, A)
----------
Set A
Set B
Set C
IsSubset(B, A)
IsSubset(C, A)
Equal(C, C)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(C, C)
----------
Set A
Set B
Set C
ℹ Operations:                                                substance synthesizer 14:07:13
 Delete IsSubset(B, A)
Delete IsSubset(C, A)
----------
Set A
Set B
Set C
IsSubset(C, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Delete IsSubset(B, A)
----------
Set A
Set B
Set C
Equal(C, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(C, A)
Delete IsSubset(C, A)
Delete IsSubset(B, A)
----------
Set A
Set B
Set C
Equal(B, C)
Equal(C, A)
ℹ Operations:                                                substance synthesizer 14:07:13
 Add Equal(B, C)
Add Equal(C, A)
Delete IsSubset(C, A)
Delete IsSubset(B, A)
----------
```

In `automator`, run `yarn start batch  registry.json artifacts --render=browser --src-prefix=progs --folders` and obtain a static site with the synthesized diagrams:
![image](https://user-images.githubusercontent.com/11740102/117336776-9127c080-ae6a-11eb-98b5-947821c42415.png)


# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

hypotext commented on packages/core/src/compiler/Substance.ts:
```diff
@@ -550,13 +549,96 @@ const checkField = (decons: Deconstructor, env: Env): ResultWithType => {
+
+//#region Substance pretty printer
+
+export const prettySubstance = (prog: SubProg): string =>
```
might be nice to pull all the AST pretty printing code (Sub + Sty + DSL) into one module, as currently my code is scattered around a few different files

wodeni replied:
Normally the pretty-printer ships with the compiler. I think it's okay to just export top-level functions in each compiler module (i.e. `Substance`, `Domain`, and `Style`).

hypotext replied:
Sounds good to me!

hypotext commented on packages/core/src/contrib/Constraints.ts:
```diff
@@ -798,7 +821,7 @@ const dsqBP = (p: any, rect: any): VarAD => {
  */
 const lerp = (l: VarAD, r: VarAD, k: VarAD): VarAD => {
-  // TODO: Rewrite the lerp code to be more concise
+  // TODO: Rewrite the lerp code to be1more concise
```
typo?

hypotext commented on packages/core/src/contrib/Constraints.ts:
```diff
@@ -683,6 +684,28 @@ export const constrDict = {
     // if x < y then 0 else (x - y)^2
     return ifCond(lt(x, y), constOf(0), squared(sub(x, y)));
   },
+
```
can you add a docstring comment, like the other constraints?

hypotext commented on packages/core/src/contrib/Constraints.ts:
```diff
@@ -683,6 +684,28 @@ export const constrDict = {
+      const v1 = ops.vsub(c1, c2);
+      const v2 = ops.vsub(c2, c3);
+      const v3 = ops.vsub(c1, c3);
+
```
can you add an inline comment about the math of this at a high level? it's not intuitive to me why this should work to make them collinear if it's operating purely on the norms of the vectors, not their angles

hypotext commented on packages/core/src/types/errors.ts:
```diff
@@ -83,7 +83,7 @@ export interface TypeNotFound {
   tag: "VarNotFound";
   variable: Identifier;
-  possibleVars?: string[]; // TODO: use Identifier type, but need to store them in env
+  possibleVars?: Identifier[]; // TODO: use Identifier type, but need to store them in env
```
remove TODO?

hypotext commented on packages/core/src/utils/Util.ts:
```diff
@@ -3,7 +3,7 @@ import { times } from "lodash";
 import { Properties } from "types/shape";
 
-seedrandom("secret-seed", { global: true }); // HACK: constant seed for pseudorandomness
+// seedrandom("secret-seed", { global: true }); // HACK: constant seed for pseudorandomness
```
do we no longer use a PRG?

hypotext commented on packages/synthesizer/__tests__/euclidean.sty:
```diff
@@ -0,0 +1,511 @@
+Colors {
```
should the tests use the `examples/` versions instead of copying the code here?

wodeni replied:
These files are specific to the examples I'm trying to do, so no. The changes to the Euclidean style are improvements that don't cause regressions, so I'm inclined to leave it in.

Existing examples are okay, the two-vectors example here:
![image](https://user-images.githubusercontent.com/11740102/119286749-b195a000-bc0a-11eb-8ad8-c5579a7b4afd.png)

hypotext reviewed (approved):
- in `automator`, following the readme, after doing this (`yarn start batch registry.json artifacts --render=browser --src-prefix=progs --folders`) and doing this (`yarn start render out browser`), I got this failure, is this expected?

I was looking at this comment. These commands aren't the same as the ones in the README though.

wodeni commented on packages/automator/stat.pug:
```diff
@@ -0,0 +1,36 @@
+doctype html
```
This is old code from [penrose-artifacts](https://github.com/penrose/penrose-artifacts)

wodeni commented on examples/geometry-domain/euclidean-simple.sty:
```diff
@@ -69,61 +69,69 @@ const {
+Point p
+with Plane P
+where In(p, P) {
+  -- TODO: the problem is that this ensures the padding is const? Or is > padding okay?
```
Nope. This is from the original Style program, not my comment.

wodeni commented:
>   * e.g. `euclidean-simple.sty`) -- can you verify that any program triple that already uses this style program works?

There's none in the registry, but I reverted my changes there anyway. I have my most recent changes in `synthesizer` already.

>   * and `twoVectorsPerp-unsugared.sub` -- does this still work?

Yep, `yarn test` produces the excepted result.
![image](https://user-images.githubusercontent.com/11740102/119287333-ece49e80-bc0b-11eb-9e87-7937656ae1cf.png)

> * I ran into some weird build problem when I tried to run the command `yarn set-example` in the directory. It was resolved after I ran `yarn`. Is this needed on installation? If so, can you add it to the README?

Whenever you have new dependencies, you'll need to run `yarn` again. I updated the README to refer to the installation page.

> * When running the `automator` command, I see in the command line that it's possible to generate failing programs. Is that intentional? Or is it a potential problem with the Style program or compiler?

It's possible and not in the scope of the current project so far. We might come back to this issue later when we are actually analyzing Style programs during synthesis. 

> * in `automator`, following the readme, after doing this (`yarn start batch registry.json artifacts --render=browser --src-prefix=progs --folders`) and doing this (`yarn start render out browser`), I got this failure, is this expected?

Not sure where you got this command from. The README instruction is asking you to run:
* `yarn start batch registry.json out/ --src-prefix=../../examples --folders`
*  `yarn start render out browser`

The first command you ran already generates the site, so your second command is unnecessary and incorrect because the first generated `artifacts`, not `out`.

> * _regarding changes in Style -- have you done a quick check to verify that any parts of the system that depend on `dummyNode` still work (or that nothing depends on that AST type)?_

All tests in the test suite are still passing. A quick visual check seems to be fine.

hypotext commented:
Ok, that all looks good, thanks!

> Not sure where you got this command from. 

I just looked at `packages/automator/README.md`, which said (at the time I ran it):

> - Run `yarn start batch registry.json out/ --src-prefix=../../examples --folders` in this directory. (...)
> - Run `yarn start render out browser` to generate a static site.

So, I followed those instructions.

wodeni commented:
> I just looked at `packages/automator/README.md`, which said (at the time I ran it):
> 
> 
> 
> > - Run `yarn start batch registry.json out/ --src-prefix=../../examples --folders` in this directory. (...)
> 
> > - Run `yarn start render out browser` to generate a static site.
> 
> 
> 
> So, I followed those instructions.

These are the right commands to run. I was looking at your earlier comment:

> - in `automator`, following the readme, after doing this (`yarn start batch registry.json artifacts --render=browser --src-prefix=progs --folders`) and doing this (`yarn start render out browser`), I got this failure, is this expected?

Replacing the first command with the one in the README should do it.
